storey/flow.py

```diff
@@ -608,73 +611,115 @@ async def _handle_completed(self, event, response):
+                self._timeout_task_key = None
+            await self._emit_batch(key)
+
+    async def _sleep_and_emit(self):
+        while self._batch:
+            key = next(iter(self._batch.keys()))
+            time_delta = datetime.now(timezone.utc) - self._batch_time[key]
+            delta_seconds = time_delta.total_seconds()
+            if delta_seconds < self._timeout_secs:
+                self._timeout_task_key = key
+                await asyncio.sleep(self._timeout_secs - delta_seconds)
+            await self._emit_batch(key)
 
-            if self._event_count == self._max_events:
-                if self._timeout_task and not self._timeout_task.cancelled():
-                    self._timeout_task.cancel()
-                await self._emit_batch()
+        self._timeout_task = None
+        self._timeout_task_key = None
+
+    def _event_to_batch_entry(self, event):
+        return self._get_event_or_body(event)
 
-    async def _emit_batch(self):
-        if len(self._batch) > 0:
-            batch_to_emit = self._batch
-            batch_time = self._batch_time
-            self._batch = []
-            self._batch_time = None
-            self._event_count = 0
+    async def _emit_batch(self, batch_key: Optional[str] = None):
+        batch_to_emit = self._batch.pop(batch_key)
+        batch_time = self._batch_time.pop(batch_key)
+        del self._event_count[batch_key]
+        await self._emit(batch_to_emit, batch_time)
 
-            await self._emit(batch_to_emit, batch_time)
+    async def _emit_all(self):
+        for key in list(self._batch.keys()):
+            await self._emit_batch(key)
 
 
 class Batch(_Batching):
     """
-    Batches events into lists of up to max_events events. Each emitted list contained max_events events, unless timeout_secs seconds
-    have passed since the first event in the batch was received, at which the batch is emitted with potentially fewer than max_events
-    event.
-    :param max_events: Maximum number of events per emitted batch. Set to None to emit all events in one batch on flow termination.
-    :type max_events: int or None
+    Batches events into lists of up to max_events events. Each emitted list contained max_events events, unless
+    timeout_secs seconds have passed since the first event in the batch was received, at which the batch is emitted with
+    potentially fewer than max_events event.
+    :param max_events: Maximum number of events per emitted batch. Set to None to emit all events in one batch on flow
+    termination.
     :param timeout_secs: Maximum number of seconds to wait before a batch is emitted.
-    :type timeout_secs: int
+    :param key: The key by which events are grouped. By default (None), events are not grouped.
+    Other options may be:
+    Set a '$key' to group events by the Event.key property.
+    set a 'str' key to group events by Event.body[str].
+    set a Callable[Union[Event, Any], Any] to group events by a a custom key extractor.
```

Comment:
Isn't `Union[Event, Any]` just `Any`? :)

Reply:
Indeed something or any == Any, but in case of `full_event=True` its specifically `Event`, you think Any is better?